### PR TITLE
make ledger_ids and currency_ids configurable

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR removes the need to duplicate ledger id specs in the skills and makes the default ledger id available on the skill context. It also makes the currency denominations available on skill context.

## Fixes

- #1819

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/aea.py [hunk 1]
```diff
@@ -89,6 +90,8 @@ def __init__(
         connection_exception_policy: ExceptionPolicyEnum = ExceptionPolicyEnum.propagate,
         loop_mode: Optional[str] = None,
         runtime_mode: Optional[str] = None,
+        default_ledger: Optional[str] = None,
```
Comment (DavidMinarsch) [inline aea/aea.py:93]: At some point we need to separate the config for the AEA from that of the context here. But this PR does not do that.

File: aea/aea.py [hunk 2]
```diff
@@ -89,6 +90,8 @@ def __init__(
...
+        ledger_id_to_currency_denom: Optional[Dict[str, str]] = None,
```
Comment (DavidMinarsch) [inline aea/aea.py:94]: Not sure about the name. Maybe just `currency_denominations`?

File: aea/aea.py [hunk 3]
```diff
@@ -148,13 +153,25 @@ def __init__(
...
 
+        default_ledger_id = (
+            default_ledger
+            if default_ledger is not None
+            else identity.default_address_key
+        )
+        ledger_id_to_currency_denom = (
+            ledger_id_to_currency_denom
+            if ledger_id_to_currency_denom is not None
+            else DEFAULT_LEDGER_ID_TO_CURRENCY_DENOM
```
Comment (DavidMinarsch) [inline aea/aea.py:164]: Maybe the default should be an empty dict

File: aea/aea_builder.py [hunk 4]
```diff
@@ -345,8 +349,9 @@ def _reset(self, is_full_reset: bool = False) -> None:
         self._build_called: bool = False
         if not is_full_reset:
             return
-        self._default_ledger = DEFAULT_LEDGER
-        self._default_connection: PublicId = DEFAULT_CONNECTION
+        self._default_ledger: Optional[str] = None
+        self._ledger_id_to_currency_denom: Dict[str, str] = {}
+        self._default_connection: Optional[PublicId] = None
```
Comment (DavidMinarsch) [inline aea/aea_builder.py:354]: This is unified so that all defaults are None/empty

File: aea/aea_builder.py [hunk 5]
```diff
@@ -1192,16 +1232,6 @@ def set_from_configuration(
...
 
-        if (
-            agent_configuration._default_connection  # pylint: disable=protected-access
-            is None
-        ):
-            self.set_default_connection(DEFAULT_CONNECTION)
-        else:
-            self.set_default_connection(
-                PublicId.from_str(agent_configuration.default_connection)
-            )
```
Comment (DavidMinarsch) [inline aea/aea_builder.py:1203]: Removed to make consistent

File: aea/cli/create.py [hunk 6]
```diff
@@ -177,9 +177,9 @@ def _crete_agent_config(ctx: Context, agent_name: str, set_author: str) -> Agent
         license_=DEFAULT_LICENSE,
         registry_path=os.path.join("..", DEFAULT_REGISTRY_PATH),
         description="",
+        default_ledger=DEFAULT_LEDGER,
+        default_connection=str(DEFAULT_CONNECTION),
     )
-    agent_config.default_connection = str(DEFAULT_CONNECTION)
-    agent_config.default_ledger = DEFAULT_LEDGER
```
Comment (DavidMinarsch) [inline aea/cli/create.py:182]: to make consistent, removed

File: aea/configurations/base.py [hunk 7]
```diff
@@ -1777,45 +1789,6 @@ def connection_private_key_paths_dict(self) -> Dict[str, str]:
             key: path for key, path in self.connection_private_key_paths.read_all()
         }
 
-    @property
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1780]: These are inconsistent relative to the other properties/args

File: aea/crypto/ethereum.py [hunk 8]
```diff
@@ -49,6 +49,7 @@
 DEFAULT_ADDRESS = "https://ropsten.infura.io/v3/f00f7b3ba0e848ddbdc8941c527447fe"
 DEFAULT_CHAIN_ID = 3
 DEFAULT_GAS_PRICE = "50"
+DEFAULT_CURRENCY_DENOM = "gwei"
```
Comment (DavidMinarsch) [inline aea/crypto/ethereum.py:52]: Check!

File: aea/identity/base.py [hunk 9]
```diff
@@ -70,6 +70,12 @@ def __init__(
         if addresses is None:
             addresses = {default_address_key: address}
         self._addresses = addresses
+        self._default_address_key = default_address_key
```
Comment (DavidMinarsch) [inline aea/identity/base.py:73]: For consistency, could be relabelled to `default_ledger_id`

File: aea/crypto/ethereum.py (hunk 8 above)
Comment (DavidMinarsch) [reply aea/crypto/ethereum.py:52]: Needs to be wei

File: aea/aea.py (hunk 2 above)
Comment (marcofavorito) [reply aea/aea.py:94]: `amount` instead of `denomination`?

Comment (marcofavorito) [reply aea/aea.py:94]: yes `currency_denominations` is preferable since it is shorter, the fact that it's a map comes from the type hint

Comment (marcofavorito) [reply aea/aea.py:94]: but both works at the end

File: aea/identity/base.py (hunk 9 above)
Comment (marcofavorito) [reply aea/identity/base.py:73]: ok

File: aea/aea.py (hunk 3 above)
Comment (marcofavorito) [reply aea/aea.py:164]: why?

File: aea/aea_builder.py (hunk 4 above)
Comment (marcofavorito) [reply aea/aea_builder.py:354]: ok

**Pull request merged**